### ISSUE TITLE
MIT 6.005 Software Construction: Test-First Programming

### ISSUE THREAD

**Issue by YvesMuyaBenda** [dwyl/learn-tdd#94]
Though the course uses Java, generally useful discussion  on testing within the larger context of writing a specification, and thinking of properties: [Test-First Programming](https://ocw.mit.edu/ans7870/6/6.005/s16/classes/03-testing/). The material gives a wider context within which to understand the specific coin-changing example being used, in that, though the tutorial is very careful to speak about the specification--requirements by any other name will still be requirements--but spends less time discussing how tests cases are chosen, with the same non-discussion of choosing test cases in the Elm-Architecture in Javascript piece, where the simplicity of the functions obscures what choice is being made in the test cases: my basic understanding is that the partitioning should follow from first the specification, and then the general properties of the functions being described, but most discussion of testing does not tend to discuss general properties.

**Comment by kfern** [comment]
Hi  @YvesMuyaBenda

For a beginner level I think it is not convenient to use a different technology like Java. It can disorient those who are following the manual.

**Comment by nelsonic** [comment]
@YvesMuyaBenda agree with @kfern, when I started writing tests back in '09 the _only_ tutorials were in Java and they were _needlessly_ complex (_to be `expected` with a class-based language_) 😉

**Comment by YvesMuyaBenda** [comment]
@nelsonic @kfern Yes, Java is a rather cumbersome langauge, with cumbersom tools, yes? My point in supplying the link was not to suggest that folk beginning in Javascript jump into the complexities of Java (that would be a needless time wasting diversion consuming large resources of time and energy and mental space), what I was referring to in the examples was the very clear discussion on testing which links thoughts of the relationship between specification and testing, testing and general properties,  testing strategy, and the cleanest discussion I have seen about how to create example based tests based upon partitioning the set of  input values, among other things. All of this can be abstracted into the context of Javascript, since these are _general_principles, as in, in choosing the test cases, what is the choice based upon? Now, these are all extra thoughts, the first step would be to be able to think of _any_ examples whatsoever. Likewise, with the link on the Python podcast, I was not suggesting that the beginning Javascripter take a diversion into python. It is just that some of the best materials explaining aspects of testing are not (in my limited experience) written in Javascript. For example, from my point of view, in F&C folk are suggested to learn to do random testing without having seen the wider context of property-based testing, which is most clearly discussed within the functional programming community [An introduction to property based testing](https://fsharpforfunandprofit.com/pbt/). From where I stand right now (and admittely I speak from my own point of view and limited experience), that would amount to rolling your own proprety based-testing framework from scratch without the wider context, by ad-hoc creating generators that create random values that conform to the types of your function without any of the thinking tools that have been created. For myself, I seek the best explanations first, for example the basic question of TDD was written in the framework of learning Ruby, and so eventually what seems to be happening is that all programming langauges are beginning to look like different types of pseudocode, since I am abstracting all the time.

To summarise, it was the principles discussed in that discussion of testing within Java and that I thought would be useful, not the Java language itself. Which is why I would never put the link within resources based on raw beginners who started with Javascript, since I would not expect them to know of or be able to abstract general principels apart from the langauge in which a tutorial is written. 

Of course, I defer to the wider community's wishes, and am happy to go along with what is decided. I appreciate the effort that it took to create DWYL.

**Comment by newswim** [comment]
There is at least one PBT library in JS that I'm aware of - https://github.com/jsverify/jsverify

A lot of the ideas come from Haskell's QuickCheck, which is extremely interesting in its own right...

<img width="855" alt="screen shot 2017-07-17 at 5 22 01 pm" src="https://user-images.githubusercontent.com/6667096/28292423-7f7c0668-6b14-11e7-91cc-1785af4a11a8.png">

See: [YouTube Link](https://youtu.be/AfaNEebCDos)

Also, AFAIK the test framework most often used in Elm works similarly, but they call it [Fuzz Testing](https://github.com/elm-community/elm-test).

**Comment by YvesMuyaBenda** [comment]
@newswim the best documented QuickCheck derivative that I know of is ScalaCheck: https://www.amazon.com/ScalaCheck-Definitive-Guide-Rickard-Nilsson/dp/0981531695/ref=sr_1_1?ie=UTF8&qid=1500406882&sr=8-1&keywords=scalacheck
